Delegator refactor

andrewvc opened the pull request:
This WIP cleans up everything related to output delegators and specifically cleans up the nomenclature WRT `concurrency`.

This is ready for review, though tests around `Pipeline#plugin` need some work. I need to write some spec helpers to let us properly test alternate registries to satisfy this. Additionally we need to make the plugin registry injectable for proper testability. I'm deliberating doing this in a separate PR.

Fixes #5661  #5658 #5659 #5754 all related to meta issue https://github.com/elastic/logstash/issues/5662

andrewvc commented on logstash-core/lib/logstash/api/commands/stats.rb:
```diff
@@ -70,7 +70,7 @@ def plugin_stats(stats, plugin_type)
             # This is to produce an array of data points, one point for each
             # plugin instance.
-            return [] unless stats[:plugins].include?(plugin_type)
+            return [] unless stats[:plugins] && stats[:plugins].include?(plugin_type)
```
Fixes an issue where this might run before `:plugins` is defined.

andrewvc commented on logstash-core/lib/logstash/filter_delegator.rb:
```diff
@@ -13,15 +13,14 @@ class FilterDelegator
-    def initialize(logger, klass, metric, *args)
-      options = args.reduce({}, :merge)
-
+    def initialize(logger, klass, metric, plugin_args)
```
Renamed options to `plugin_args` to be more clear and consistent across various classes.

guyboertje commented on logstash-core/lib/logstash/output_delegator.rb:
```diff
@@ -1,192 +1,60 @@
-  # Needed for testing, so private
-  attr_reader :threadsafe_worker, :worker_queue
-end end
+end; end
```
Looking at this class holistically...
- Method `id` and ivar `@id`. Consider only using the `id` method and provide nil fallback `@plugin_args.fetch('id', SecureRandom.hex(2))

andrewvc replied:
Good point on only using the ID method. In my next commit I'll actually require the ID be set on plugin create. It's a sane requirement and reduces mystery.

Hmmm, @jordansissel and I arrived at the term `@strategy` after a discussion. The thing is that at some point you an `OutputDelegator` and an `OutputProxy/OutputDelegate` AND an `Output` and that all gets very confusing in terms of hierarchy.

As is you say "We have an output delegator, it has a strategy, the strategy controls the output itself", which, to my ears is a bit easier to understand.

guyboertje commented on logstash-core/lib/logstash/output_delegator.rb:
```diff
@@ -1,192 +1,60 @@
-
-    @events_received = Concurrent::AtomicFixnum.new(0)
+    @id = plugin_args["id"]
+    @namespaced_metric = metric.namespace("#{@output_class.config_name}_#{@id}".to_sym)
```
Consider using the method `config_name` for the metric namespace, further more that would remove the need to create an ivar on line 17.

andrewvc replied:
I've actually refactored this to use `#id` since it should be unique :)

guyboertje commented on logstash-core/lib/logstash/output_delegator.rb:
```diff
@@ -1,192 +1,60 @@
+  def build_strategy!
+    # TODO: Make this injected
+    @strategy_registry = LogStash::OutputDelegatorStrategyRegistry.instance
+    @strategy_class = @strategy_registry.get(self.concurrency)
```
Consider raising the "Could not find output delegator strategy of type..." error in the `instance.get()` method (it is probably where it belongs, as any other method using it will have to also test and raise).
However, maybe `class_for` makes more sense .e.g. `.instance.class_for(self.concurrency).new(...)`
Gives...
```ruby
  def build_strategy!
    @output_delegate = LogStash::OutputDelegatorStrategyRegistry.instance
        .class_for(self.concurrency)
        .new(@logger, @output_class, @metric, @xopts, @plugin_args)
  end
```

guyboertje commented on logstash-core/lib/logstash/output_delegator_strategies/legacy_worker.rb:
```diff
@@ -0,0 +1,27 @@
+  
+  def multi_receive(events)
+    worker = @worker_queue.pop
+  ensure
```
Did you mean to not do anything with events? To be consistent with A) th previous code and B) the other strategies, consider this:
```ruby
  def multi_receive(events)
    output = @worker_queue.pop
    begin
      output.multi_receive(events)
    ensure
      @worker_queue.push(output)
    end
  end
```

jordansissel replied:
Agreed. Is legacy missing a complete implementation?

andrewvc replied:
That is totally a bug, and I'm clearly missing a test to boot. Good catch :)

guyboertje commented on logstash-core/lib/logstash/output_delegator_strategy_registry.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+  
+  def get(type)
+    @map[type]
```
As mentioned above, consider changing this name to `class_for(type)` (its not immediately obvious from the class name `OutputDelegatorStrategyRegistry` that classes are stored within) and raising an error when not found here.

jordansissel commented on logstash-core/lib/logstash/output_delegator.rb:
```diff
@@ -1,192 +1,60 @@
 
   def config_name
-    @klass.config_name
+    "#{@output_class.config_name}_#{id}"
```
Apologies if this is inarticulate, but this feels wrong. The `config_name` is like the name of the plugin, right? For this delegator, it becomes `name-id` where `id` is a computed value.

Can we use a better name than `config_name` for this method -- something that better fits the purpose? Maybe `plugin_id` or something?

guyboertje replied:
I feel the same way, however I did not comment on this because it is part of the public API. But, if id is now an intrinsic part of the config_name then it is correct.

andrewvc replied:
Thanks for the good feedback. Looking at this closer I suspect we no longer need this method, I'm going to try to remove it and see if there are any issues there.

andrewvc replied:
OK, so it looks like this should just proxy through to @output_class.config_name, and in other places where we use uniqueness #id should be invoked directly. Will push this up in my next commit.

jordansissel commented on logstash-core/lib/logstash/output_delegator_strategy_registry.rb:
```diff
@@ -0,0 +1,28 @@
+  # This is generally used as a singleton
+  # Except perhaps during testing
+  def self.instance
+    @instance ||= self.new
```
No change required, but ruby has a `Singleton` library in stdlib that does this.

andrewvc replied:
I'm not using singleton here because during testing I do want the ability to create a new instance without the default plugin list. I think we need a better name for this pattern (do you know of one that is already in use?) as `instance` too strongly implies a singleton. 

This is more like 'it works as a singleton by default but it is totally legit to make another instance, which is very useful for testing.'

andrewvc replied:
OK, so it looks like this should just proxy through to `@output_class.config_name`, and in other places where we use uniqueness `#id` should be invoked directly. Will push this up in my next commit.

jordansissel commented on logstash-core/lib/logstash/outputs/base.rb:
```diff
@@ -20,42 +20,40 @@ class LogStash::Outputs::Base < LogStash::Plugin
+      # TODO remove this in Logstash 6.0
+      # when we no longer support the :legacy_worker type
+      # This is hacky, but it can only be here
+      if type == :legacy_worker
```
Feels inconsistent to have them called `shared`, `single`, and then a two-word name `legacy worker`. Can we call it `legacy` instead of `legacy_worker`? What's the story on the `_worker` suffix?

andrewvc replied:
Fair point, I'm cool calling it `legacy`. The `worker` suffix was meant to inform the user that this is the one that goes with the `workers` setting. The other ones no longer support that setting.

jordansissel commented on docs/static/include/pluginbody.asciidoc:
```diff
@@ -275,40 +275,29 @@ require "logstash/namespace"
+  # This is the default concurrency level of this output plugin.
+  # You do not need to explicitly set it.
+  # When configured as :single a single instance of the Output will be shared among the
+  # pipeline threads. Access to the `#multi_receive/#receive` method will be synchronized
```
should we call this `pipeline workers` instead of `pipeline threads` ?

andrewvc replied:
How bout I rename it pipeline worker threads. Since we're talking about synchronization I like that we're informing the reader that these are os threads.

jordansissel commented on docs/static/include/pluginbody.asciidoc:
```diff
@@ -275,40 +275,29 @@ require "logstash/namespace"
+  # You do not need to explicitly set it.
+  # When configured as :single a single instance of the Output will be shared among the
+  # pipeline threads. Access to the `#multi_receive/#receive` method will be synchronized
+  # i.e. only one thread will be active at a time making threadsafety much simpler.
```
Similar here, thread vs worker terms.

jordansissel commented:
Overall, I am big +1 on the idea here. My comments are highly localized and mostly around "is this the best term for this thing?" or similar descriptive concerns.

I haven't reviewed the test changes, yet, but I do see they are passing in travis.

guyboertje commented on logstash-core/lib/logstash/output_delegator_strategies/legacy_worker.rb:
```diff
@@ -0,0 +1,27 @@
+# Remove this in Logstash 6.0
+module LogStash module OutputDelegatorStrategies class LegacyWorker
+  ::LogStash::OutputDelegatorStrategyRegistry.instance.register(:legacy_worker, self)
```
I am concerned about this register statement.

In practice this is not a problem because we are in full control of how and when these classes are retrieved. That said, the current code would give the casual reader the idea that this is OK but there is a bug lurking in it. The class is registered without methods - they are added shortly after registration. However, if the class is retrieved and instantiated before the methods are added the instance will have no public API.

Consider moving the registration to after the methods are added. Alternatives:
```ruby
module LogStash
  module OutputDelegatorStrategies class LegacyWorker
    # rest of class definition omitted for illustration
  end; end

  OutputDelegatorStrategyRegistry.instance.register(
    :legacy_worker, OutputDelegatorStrategies::LegacyWorker)
end
```
or
```ruby
module LogStash module OutputDelegatorStrategies class LegacyWorker
  # rest of class definition omitted for illustration
  
  ::LogStash::OutputDelegatorStrategyRegistry.instance.register(:legacy_worker, self)
end; end; end
```

guyboertje commented on logstash-core/lib/logstash/pipeline.rb:
```diff
@@ -45,11 +46,21 @@ module LogStash; class Pipeline
     @reporter = LogStash::PipelineReporter.new(@logger, self)
 
+    # Maybe make this configurable via an option at some point
+    @output_delegator_strategy_registry = ::LogStash::OutputDelegatorStrategyRegistry.instance
```
This reference does not seem to be used anywhere.

andrewvc replied:
Good point. My original intent was to pass this as an argument to the output delegator. I think I'm going to go ahead and make that refactor now as its worth it. 

Ultimately this should simply be the default value, with the calling code able to override it so we can do richer testing.

guyboertje commented on logstash-core/lib/logstash/pipeline.rb:
```diff
@@ -387,23 +398,38 @@ def shutdown_workers
+               new_plugin = klass.new(args)
+               namespaced_metric = type_scoped_metric.namespace(id)
+               new_plugin.metric = namespaced_metric
+               new_plugin
```
Consider renaming the `new_plugin` variable to the self documenting`input`.

guyboertje commented on logstash-core/lib/logstash/plugin.rb:
```diff
@@ -45,8 +45,11 @@ def eql?(other)
     @params = LogStash::Util.deep_clone(params)
+    # The id should always be defined normally, but in tests that might not be the case
+    # In the future we may make this more strict in the Plugin API
+    @params["id"] ||= params["id"] || "#{self.class.config_name}_#{SecureRandom.uuid}"
```
Do you need the `params["id"] || `?

andrewvc replied:
I don't believe so! Good catch

guyboertje commented:
This is looking good. Just a few considerations to address and its not WIP anymore.

guyboertje commented on logstash-core/lib/logstash/output_delegator_strategy_registry.rb:
```diff
@@ -17,8 +19,14 @@ def types
+    klass = @map[type]
+
+    if !klass
+      raise InvalidStrategyError, "Could not find output delegator strategy of type '#{type}'. Valid strategies: #{@strategy_registry.types}"
```
Nice!

andrewvc commented:
@guyboertje @jordansissel I believe I've addressed all your comments. I've tried to separate things into separate commits to make viewing the changes I made today more visible. I'll squash them all before final merge.

guyboertje commented:
LGTM

jordansissel commented on logstash-core/lib/logstash/pipeline.rb:
```diff
@@ -32,24 +32,34 @@ module LogStash; class Pipeline
     :filter_queue_client,
-    :input_queue_client
+    :input_queue_client,
+    :xopts
```
the `xopts` name is confusing to me. What does it mean? Can we use a word or two to describe `word1_word2` or something instead of this `xopts` name?

andrewvc replied:
The basic idea was to provide extended options. Do you prefer `extended_options`? I dislike `options` because we already have settings, but `settings` unfortunately is for user configurable things and not rich objects ATM.

andrewvc replied:
How about `xoptions`?

jordansissel commented:
General LGTM. I'd like to see `xopts` renamed to something more descriptive for me, but we can change that later if desired.
